# -*- coding: utf-8 -*-
# Generated by Django 1.11 on 2018-04-17 23:53
from __future__ import unicode_literals

from django.db import migrations, models


class Migration(migrations.Migration):

    dependencies = [
        ('seqr', '0035_auto_20180417_2132'),
    ]

    operations = [
        migrations.RenameField(
            model_name='variantfunctionaldata',
            old_name='search_url',
            new_name='search_parameters',
        ),
        migrations.AddField(
            model_name='variantnote',
            name='submit_to_clinvar',
            field=models.BooleanField(default=False),
        ),
        migrations.AlterField(
            model_name='variantfunctionaldata',
            name='functional_data_tag',
            field=models.TextField(choices=[(b'Functional Data', ((b'Biochemical Function', b'{"color": "#311B92", "metadata_input_props": "maxlength=\\"50\\" style=\\"width:400px\\"", "metadata_title": "Notes", "description": "Gene product performs a biochemical function shared with other known genes in the disease of interest, or consistent with the phenotype."}'), (b'Protein Interaction', b'{"color": "#4A148C", "description": "Gene product interacts with proteins previously implicated (genetically or biochemically) in the disease of interest."}'), (b'Expression', b'{"color": "#7C4DFF", "description": "Gene is expressed in tissues relevant to the disease of interest and/or is altered in expression in patients who have the disease."}'), (b'Patient Cells', b'{"color": "#B388FF", "description": "Gene and/or gene product function is demonstrably altered in patients carrying candidate mutations."}'), (b'Non-patient cells', b'{"color": "#9575CD", "description": "Gene and/or gene product function is demonstrably altered in human cell culture models carrying candidate mutations."}'), (b'Animal Model', b'{"color": "#AA00FF", "description": "Non-human animal models with a similarly disrupted copy of the affected gene show a phenotype consistent with human disease state."}'), (b'Non-human cell culture model', b'{"color": "#BA68C8", "description": "Non-human cell-culture models with a similarly disrupted copy of the affected gene show a phenotype consistent with human disease state."}'), (b'Rescue', b'{"color": "#663399", "description": "The cellular phenotype in patient-derived cells or engineered equivalents can be rescued by addition of the wild-type gene product."}'))), (b'Functional Scores', ((b'Genome-wide Linkage', b'{"color": "#880E4F", "metadata_input_props": "type=\\"number\\" required=\\"true\\" style=\\"width:120px;\\"", "metadata_title": "LOD Score", "description": "Max LOD score used in analysis to restrict where you looked for causal variants; provide best score available, whether it be a cumulative LOD score across multiple families or just the best family\'s LOD score."}'), (b'Bonferroni corrected p-value', b'{"color": "#E91E63", "metadata_input_props": "type=\\"number\\" required=\\"true\\" style=\\"width:120px;\\"", "metadata_title": "P-value", "description": "Bonferroni-corrected p-value for gene if association testing/burden testing/etc was used to identify the gene."}'), (b'Kindreds w/ Overlapping SV & Similar Phenotype', b'{"color": "#FF5252", "metadata_input_props": "type=\\"number\\" required=\\"true\\" style=\\"width:120px;\\"", "metadata_title": "#", "description": "Number of kindreds (1+) previously reported/in databases as having structural variant overlapping the gene and a similar phenotype."}'))), (b'Additional Kindreds (Literature, MME)', ((b'Additional Unrelated Kindreds w/ Causal Variants in Gene', b'{"color": "#D84315", "metadata_input_props": "type=\\"number\\" required=\\"true\\" style=\\"width:120px;\\"", "metadata_title": "# additional families", "description": "Number of additional kindreds with causal variants in this gene (Any other kindreds from collaborators, MME, literature etc). Do not count your family in this total."}'),))]),
        ),
    ]
